test/test-url.cpp

```diff
@@ -1154,3 +1154,34 @@ TEST(PunycodeTest, Safe)
     std::string encoded("http://safe.segments/");
     EXPECT_EQ(encoded, Url::Url(unencoded).punycode().str());
 }
+
+TEST(PunycodeTest, EmptySegment)
+{
+    ASSERT_THROW(Url::Url("http://foo..com/").punycode(), std::invalid_argument);
+}
+
+TEST(PunycodeTest, TrailingEmptySegment)
+{
+    ASSERT_THROW(Url::Url("http://foo../").punycode(), std::invalid_argument);
```

Comment:
It is interpreted as the `.` being the `host`, and does indeed complain about an empty segment (the leading segment).